refactor(semantic): consolidate diagnostics into single module

camc314 opened the pull request:
Diagnostic functions for semantic were previously spread across 3x files, with `diagnostics.rs` only containing a single diagnostic. This PR refactors it to have all diagnostics in a single file. When diagnostics are needded, they are referenced via `diagnostic::my_diagnostic` to be clear and precise as to where the diagnostic is coming from.

I **think** this is better than how it was before - but if you disagree Boshen, feel free to close.

camc314 commented:
* **#17490** <a href="https://app.graphite.com/github/pr/oxc-project/oxc/17490?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/oxc-project/oxc/17490?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=17490).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Boshen commented:
### Merge activity

* **Dec 30, 12:11 PM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc) once it meets the requirements.

camc314 committed:
refactor(semantic): consolidate diagnostics into single module (#17490)

Diagnostic functions for semantic were previously spread across 3x files, with `diagnostics.rs` only containing a single diagnostic. This PR refactors it to have all diagnostics in a single file. When diagnostics are needded, they are referenced via `diagnostic::my_diagnostic` to be clear and precise as to where the diagnostic is coming from.

I **think** this is better than how it was before - but if you disagree Boshen, feel free to close.
